### ISSUE TITLE
grep not getting colours enabled on OS X/BSD

### ISSUE THREAD

**Issue by daenney** [fish-shell/fish-shell#2746]
Fish ships with the following [`grep` function](https://github.com/fish-shell/fish-shell/blob/190cac07a3ff7cd2c65ac24b600808e7c1674412/share/functions/grep.fish):

```
#
# Match colors for grep, if supported
#

if command grep --color=auto --help 1>/dev/null 2>/dev/null
	function grep
		command grep --color=auto $argv
	end
end
```

However, this doesn't work for OS X even though the `grep` in OS X supports `--color=auto`. This appears to happen because `--help` exist with an exit code of 1 on BSD grep but with 0 on GNU grep.

Looking at the BSD manpage:

> EXIT STATUS
>     The grep utility exits with one of the following values:
>
>     0     One or more lines were selected.
>     1     No lines were selected.
>     >1    An error occurred.

GNU says the same thing but from the observed behaviour it looks like they special case `--help`:
> EXIT STATUS
>       The exit status is 0 if selected lines are found, and 1 if not found.  If an error occurred the exit status is 2.  (Note: POSIX error handling code should check for '2' or greater.)

Arguably GNU's behaviour is more correct since calling `--help` would never cause it to match lines to begin with. However, it would be nice to have it correctly work on both.

**Comment by bucaran** [comment]
Maybe something like this:

```fish
if command grep --help ^| command grep --quiet -- --color
    function grep
        command grep --color=auto $argv
    end
end
```

**Comment by faho** [comment]
Fixed via 613739ba072deffb7c0447de02c4a5a3ce296add. Thanks!
